File: 242 Valid Anagram/Python/Solution.py
from typing import Dict, Self


class Solution:
    def isAnagram(self: Self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        map: Dict[str, int] = {}
        for c in s:
            if c in map:
                map[c] += 1
            else:
                map[c] = 1
        
        for c in t:
            if c not in map or map[c] == 0:
                return False
            map[c] -= 1
        
        return True

File: 242 Valid Anagram/Java/Solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public boolean isAnagram(String s, String t) {
        int sLen = s.length();
        int tLen = t.length();

        if (sLen != tLen) {
            return false;
        }

        Map<Character, Integer> map = new HashMap<>();

        for (int i = 0; i < sLen; i++) {
            char c = s.charAt(i);
            if (map.containsKey(c)) {
                map.put(c, map.get(c) + 1);
            }
            else {
                map.put(c, 1);
            }
        }

        for (int i = 0; i < tLen; i++) {
            char c = t.charAt(i);
            if (!map.containsKey(c) || map.get(c) == 0) {
                return false;
            }
            map.put(c, map.get(c) - 1);
        }

        return true;
    }
}